Allow ovs-vswitchd logging by separating flow restoration

elvgarrui opened the pull request:
We used --detach on ovs-vswitchd execution to allow for flow
restoration, but this had the side effect of not having its console logs
available. In order to get those back, we use --log-file and then tail
that file in the end of the script to be able to see these logs from the
oc log command.

Jira: [OSPRH-11228](https://issues.redhat.com//browse/OSPRH-11228)

elvgarrui commented on templates/ovncontroller/bin/start-vswitchd-extras.sh:
```diff
@@ -0,0 +1,40 @@
+cleanup_flows_backup
+
+# Now, inform vswitchd that we are done.
+ovs-vsctl remove open_vswitch . other_config flow-restore-wait
```
ack, I think you are right. I'm looking into how to solve this.

averdagu commented:
/ok-to-test

karelyatin commented on templates/ovncontroller/bin/start-vswitchd.sh:
```diff
@@ -30,7 +30,14 @@ ovs-vsctl --no-wait set open_vswitch . other_config:flow-restore-wait=true
 
 # It's safe to start vswitchd now. Do it.
 # --detach to allow the execution to continue to restoring the flows.
```
can remove the --detach comment now

karelyatin commented on templates/ovncontroller/bin/execute-vswitchd.sh:
```diff
@@ -0,0 +1,20 @@
+# License for the specific language governing permissions and limitations
+# under the License.
+
+source $(dirname $0)/functions
```
source shouldn't be needed as nothing being used from there.

karelyatin commented on templates/ovncontroller/bin/start-vswitchd.sh:
```diff
@@ -30,7 +30,14 @@ ovs-vsctl --no-wait set open_vswitch . other_config:flow-restore-wait=true
 # It's safe to start vswitchd now. Do it.
 # --detach to allow the execution to continue to restoring the flows.
-/usr/sbin/ovs-vswitchd --pidfile --mlockall --detach
+$EXECUTE_VSWITCHD_SCRIPT > /dev/stdout 2>&1 &
```
why we need additonal bash script for just this command? running /usr/sbin/ovs-vswitchd --pidfile --mlockall > /dev/stdout 2>&1 & have some issues?

elvgarrui commented:
I addressed all the comments now. However, I realized that there is a chance that the pod makes 1 restart for it to work. For me it only happens when I execute the ovn-controller with my patch for the first time, any "delete pod"s done after that don't seem to generate this restart once the controller is running.

The error I see happens when executing the restore flows script.  `eval "$(cat $FLOWS_RESTORE_SCRIPT)"`
and is:
```
++ ovs-ofctl -O OpenFlow14 add-tlv-map br-int '{class=0x102,type=0x80,len=4}->tun_metadata0'
2025-04-04T10:27:24Z|00045|connmgr|INFO|br-int<->unix#11: sending NXTTMFC_ALREADY_MAPPED error reply to NXT_TLV_TABLE_MOD message
OFPT_ERROR (OF1.4) (xid=0x2): NXTTMFC_ALREADY_MAPPED
NXT_TLV_TABLE_MOD (OF1.4) (xid=0x2):
 ADD mapping table:
  class  type  length  match field
 ------  ----  ------  --------------
  0x102  0x80       4  tun_metadata0
```
I honestly have been trying to find out how this is originated (Maybe ovs-vswitchd starts too quickly?), and also I think this might not fail if we didn't have the -ex. I'm not sure if the "ALREADY_MAPPED" error should stop the script from running, but I would appreciate someone else opinion on this!

Full logs: https://paste.opendev.org/show/b1Rzox0URO1Y8boDcLZE/

elvgarrui commented:
Seems like --detach was taking into account completing the initial setup, etc...
We don't do it currently, so depending on the time of sleep that we set in
```
 if [ ! -f /var/run/openvswitch/ovs-vswitchd.pid ]; then   
    sleep 0.2                                              
 fi                                                        
```
and also depending on the time ovs takes to setup, we might get or not an error because of duplicates or missing items.

I need to find a logical solution to this, it would be better to not have a sleep here.

elvgarrui commented:
After talking with OVN folks, it seems like the error happens because ovn-controller starts configuring OVS without taking into account the flow-restore-wait flag. I'm trying to find a workaround for this on the meanwhile.
OVN issue: https://issues.redhat.com/browse/FDP-1292

karelyatin commented:
> After talking with OVN folks, it seems like the error happens because ovn-controller starts configuring OVS without taking into account the flow-restore-wait flag. I'm trying to find a workaround for this on the meanwhile. OVN issue: https://issues.redhat.com/browse/FDP-1292

@elvgarrui to confirm the issue you mentioned only happens with your PR and not with current main branch, right? to see if we can unblock this if it's also seen in main.

elvgarrui commented:
@karelyatin I have not seen the restart on main branch, so I asume with the --detach flag this is correctly handled.

karelyatin commented:
> @karelyatin I have not seen the restart on main branch, so I asume with the --detach flag this is correctly handled.

ok thx for confirming, wdyt to try alternative i.e keep --detach and include --log-file option to also log to file, and instead of sleep infinity tail -f /var/log/openvswitch/ovs-vswitchd.log ? or create a side car container just for logging tail -f /var/log/openvswitch/ovs-vswitchd.log ?

elvgarrui commented:
> ok thx for confirming, wdyt to try alternative i.e keep --detach and include --log-file option to also log to file, and instead of sleep infinity tail -f /var/log/openvswitch/ovs-vswitchd.log ? or create a side car container just for logging tail -f /var/log/openvswitch/ovs-vswitchd.log ?

I can try this for sure!! It looks easier to implement than my approach. 

Still, I have added mine to the PR as a separate commit for you to check, I'd like to know whether it is correct or not (even if I change it to the other way). I added these changes in a separate commit so that I can reset the branch and implement your approach more smoothly :)

elvgarrui commented:
/test ovn-operator-build-deploy-kuttl

elvgarrui commented:
/cherry-pick 18.0-fr2
